### DISCUSSION TITLE
CS8618 rised when it shouldn't.

### DISCUSSION THREAD

**Discussion by KevinValmo** [dotnet/roslyn#61770] [General] [answered]
Hello,

I made a [gist](https://gist.github.com/KevinValmo/4ba6a309b28f39b3ee9488ba11a4388b) with the following code: 👇

```cs
public class Foo
{
    public string Name { get; set; }

    public Foo(Action<Foo> options) // CS8618
    {
        options(this);
    }
}
public class Bar
{
    public Foo MyProperty { get; set; } = new Foo(v => v.Name = "");
}
```

I guess Roslyn is rising a warning that it should not rise:
`CS8618: Non-nullable property 'Name' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.`

IMO it should not rise this warning couse we are using the `Action<Foo>` directly in the ctor to instantiate the `Foo` object that has the string set as not nullable (couse we do not have the nullable '?').

Same happen with this 👇 snippet:

```cs
public class Foo
{
    public string Name { get; set; }

    public Foo() // CS8618
    {
        Init();
    }

    private void Init()
    {
        Name = "";
    }
}
```

Instead this 👇 wont rise CS8618 but CS8601 (I guess that's correct):

```cs
public class Foo
{
    // notice we don't have a prop but a field
    public string name;

    public Foo()
    {
        Init(ref name); // CS8601
    }

    private void Init(ref string str)
    {
        name = "";
    }
}
```

`CS8601: Possible null reference assignment.`

I'm confused about it, don't you? 
Am I missing something?

**Comment by PathogenDavid** [comment, marked as answer]
> Am I missing something?

The nullable flow analysis does not cross method boundaries. It doesn't see what's happening in your `options` lambda or your `Init` methods. This is by design.

That being said, there are workarounds for all three of these.

Addressing your examples in order:

> lambda `options` initializing `Name`

This situation would still be a problem even if the compiler could do flow analysis through multiple methods because the compiler doesn't have a way of knowing whether the caller of the constructor passed a lambda which will actually initialize `Name`.

My recommendation would be to validate `options` actually initialized `Name` using [an exception](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAYhgFcAbUiYU1AAlRkuoFgAoAAQCYBGV1tgZhqcaAMQD2Y1gG9WNOUMFsuABhoA5CGFpSaAc1QIA3DQDOB4wF9eLeQtESAFEo4AecWIB8NMQAcEUMRgTAEpZeRkbW3lff0CTBwQACygQwzComnSo7BoHDS0aFJoSclDIjNsk9DEAd2LUOoBBdF1iLRgEAFEADwBjVD8AmAcAEgAiAAMpBi0xLAcYoZCLGgATKFXisQQaH2qfDFIcQpgofwhSKAAvWimZ1Dm8zVRgiwmxuGLnx8W44OC0uUaFYWCCBEIODQAEIQdDSLLg9w0ACyOAACvsMAhjjp9EZTOZgTQALz1OruBzIEleZAAOnytFJYzGANYFiAA===) or [`Debug.Assert`](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAYhgFcAbUiYU1AAlRkuoFgAoAAQCYBGVtgBhpsuAOgAiUCAHMYAewDOCKAGM5Abla8AzII40AYjJmsA3qxrnB2oQIByEMLWM1JqBKppzX7gL4aWFy31DAAohDgAeAxkAPhoZAAdFGRg5AEozC1N/AIsEpJTghAALKDT1bJyaUVRgYklhAEE5T3QEYLsHGlKaWQQeslJU8oDfFlG2K10AIQh0EwzzCaCZGgBZHAAFdASMBBwaJxc3Dy8abxoAXh7UAHdl4ORL2ORhDtorgCIPodZvIA). Both will suppress the warning.

> `Name` property initialized by `Init` method

[Mark `Init` with `MemberNotNullAttribute`](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAYhgFcAbUiYU1AAlRkuoFgAoAAQAYa2BGAOgAiUCAHMYAewDOCKAGNJfAMLiAJqgCCDUjklRJAblas2AZm4AmGgDFx41gG9WNZ9zO8uAOQhha9miNQEfRpJQOCAXyMWF1drWwAKAEonF0domJcASRgoBCTDdOdIlhTnAG0AWVQwYAwPcQQPMlJ4hh9xLHivH0TEgF1SmgAHTGQIBFo2ABYabNykwbSMl27aAF4aACJNgpji8KA) to tell the compiler that `Init` will ensure `Name` is not null.

> Possible null reference assignment when passing `name` field to `Init`

The easiest solution here is to [change the parameter from `ref` to `out`](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAYhgFcAbUiYU1AAlRkuoFgAoVgAQGYb2AmGgGIB7IawDerGlJoB6GTRhCEUAMa0A7rQAmQmAHIENABYQ0NCDQAO6IZZrBihi1iipSWydK48AjAAYFCDBUAG5WTylvYSEACgBKCJoJFmlUmgBJGCgEGKFHQOC4sJTpAF9wkqlrKGQIBFp2ABYMrJy8w3Z/AtQEyqTE1IZgmgBeGgAiceLU8pZSoA===).

Alternatively, if `ref` is desirable here you can mark the parameter [as nullable and with `NotNullAttribute`](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAYhgFcAbUiYU1AAlRkuoFgAoAAQAYa2BGAOgAiUCAHMYAewDOCKAGNJfAMLiAJqgCCDUjklRJAblas2AZm4AmGgDFx41gG9WNZzQD0rmhJmzaAd1oq4jAA5Ag0ABYQaDQQNAAO6OJxNMDEYbFYUKikKk4uptw8XAxgqIYsec4FNuIAFACUlTSOLC5tNACSMFAIteioWJ4QpfXlbQC+Rq0uCVDIEAi0bAAsnd29ANoAcuIIW2SkALo0/YO8HAD8QyNNLe0uJbQAvDQARK9jLpMs40A=). (This tells the compiler that the referenced input can be null when called but it won't be after the method completes.)

----------

You can read more about `MemberNotNullAttribute` and `NotNullAttribute` here: [Attributes for null-state static analysis interpreted by the C# compiler](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis)

**Reply by KevinValmo** [reply]
Thank a lot for the explanation.
Very helpful!
